### ISSUE TITLE
Features of groups

### ISSUE THREAD

**Issue by CodeWithShreyans** [gioxx/MarvellousSuspender#133]
Please complete the following information when submitting a feature request or bug report.
 * Extension version: v7.1.6.2
 * Browser name & version: Google Chrome Canary 94.0.4583.2
 * Operating system & version: Windows 11 22000.100

Add the same features present now but just for groups instead, like Unsuspend/suspend all tabs in group, Unsuspend/suspend all tabs not in group, Never suspend this group

**Comment by arthurgbranco** [comment]
Also a "Never suspend tabs in groups" to settings.

**Comment by CodeWithShreyans** [comment]
Any progress on this @gioxx?

**Comment by gioxx** [comment]
@DestroyerXyz no, take a look here: https://github.com/gioxx/MarvellousSuspender/issues/197
I can't think to maintain / introduce enhancements in TMS wihout know nothing about his future survival (cause Google Manifest V3).

**Comment by gioxx** [comment]
To be associated with #76 (once the next version is released, it will be possible to manage Chrome's native Tab Groups, and so I think we can think about including this new feature).

**Comment by gioxx** [comment]
Revisiting this with fresh eyes: interpreting the ask as per-tab-group suspension settings, being able to set a suspend timeout (or "never suspend") that applies only to tabs within a specific Tab Group, without touching global settings or other groups/ungrouped tabs.

That's a real, coherent feature, but a fair chunk of work: it needs a new per-group settings store (keyed by group, Chrome's tab group IDs aren't guaranteed stable across restarts, so matching by group title/color is more reliable but has its own edge cases), UI to set/edit those overrides (likely from the group's own right-click menu or a dedicated settings section), and suspend-logic changes to check the group override before falling back to the global timeout. Medium-to-large effort, not a quick toggle.

Given the scope, targeting this for the longer-term 10.0.0 milestone rather than one of the nearer-term ones.

**Comment by gioxx** [comment]
Update: the per-group settings work turned out more tractable than the 10.0.0 scoping suggested, the implementation is essentially ready. Moving this to 9.0.3.

**Comment by gioxx** [comment]
"Suspend/unsuspend all tabs in this group" has landed in the 9.0.3 development branch (right-click and tab-strip context menu, plus keyboard shortcuts); not released yet, still pending the next Chrome Web Store update.

Before building the remaining two asks (all tabs NOT in a group, never suspend this group): could you share a concrete use case for either one? With group-scoped suspend/unsuspend now available, we want to understand why those two would still be needed rather than assuming. Happy to build them if there's a real workflow behind it.

**Comment by jadshaker** [comment]
For the "all tabs NOT in a group", I think they might be sometimes helpful especially if pinned tabs are excluded. My workflow:
- pinned tabs : daily uses
- grouped tabs : related tabs I access frequently, but not constantly
- scattered tabs : miscellaneous stuff I need, but not right now... but I don't want them to sleep in the bookmark bar
Having the "suspend all tabs not in a group" would be useful.

For the "never suspend this group", I don't think it is as useful as the first one, but it could be useful if you have monitoring tabs or live metrics dashboards in a group ... for me personally i won't use it, but I can understand having it...

I'd be happy to work on these 2 tasks, @gioxx what do you think ?

**Comment by gioxx** [comment]
Thanks @jadshaker, that's a clear enough picture to move on both.

On the pinned-tab concern for "suspend all ungrouped tabs": already covered. There's a "Never suspend pinned tabs" option (on by default), and a bulk suspend action goes through the same eligibility checks, so pinned tabs are skipped automatically while that option is on. No special handling needed in the new action.

On "never suspend this group": the monitoring/dashboard case is reason enough to build it. Heads-up on scope: there's no per-group settings store in TMS today (only the group-scoped suspend/unsuspend action that landed for the next release), so this one needs a small persistent store keyed by group. Chrome's numeric group IDs aren't stable across restarts, so matching on group title + color is the more reliable key.

Green light on both, and thanks for offering to take them, this would be your first contribution here and it's welcome. A separate PR per ask is easiest to review. You can mirror the pattern from the group suspend/unsuspend action in #438 for the ungrouped one.

Targeting 9.1.0: 9.0.3 is being wrapped for release right now, so these land in the version right after.

**Comment by jadshaker** [comment]
Hi @gioxx 

I have just finished working on the "suspend / unsuspend all tabs not in a group" thing. I'll start working on the other PR in a bit.

**Comment by jadshaker** [comment]
regarding the other feature, we can go with group title + color as you mentioned

> Chrome's numeric group IDs aren't stable across restarts, so matching on group title + color is the more reliable key.

but this might be unreliable in cases where people have many unnamed groups, which im not sure if some body does often or not. We might need to go into very complex directions in that case; however, another much easier solution would be something like this:

```
onNewTabGroup:
  key = group_title + color
  if key exists in keys:
    other_key = key + unique_urls_of_group
    key = key + unique_urls_of_group

unique_urls_of_group:
  option 1: return base url of all tabs in the group
  option 2: return full url of all tabs in the group
  option 3: return full url of the first tab + nb of tabs
```

this way, we would try to avoid conflicts as much as possible without overcomplicating the algorithm... what do you think?

For the `unique_urls_of_group` we can go with `all` tabs or `some` tabs, but again, i think this would be overcomplicating stuff, and instead we can just go with `option 3`. But they all have one problem, which is we would have listeners on every tab inside the group and on the group itself... Maybe fixing the number of tabs in this function would decrease this number of listeners, something like base/full url of the first 3 groups ...

I'm not sure regarding that, what do you think @gioxx ?

**Comment by AmineI** [comment]
> Also a "Never suspend tabs in groups" to settings.

Looking forward to the "Never suspend tabs in any tab group" setting as well 🙂.
The use case is for syncing tab groups between devices - a suspended tab doesn't sync at all unfortunately.

**Comment by gioxx** [comment]
@jadshaker thanks, and #483 looks good, I'll review it properly shortly.

I checked all of this against the codebase before answering. Summary: the global toggle is cheap, the per-group override is a moderate but self-contained follow-up, and the composite-key approach is not worth it.

**Global "never suspend tabs in any group" first, as its own small PR.** `tab.groupId` is already on the tab object everywhere `checkTabEligibilityForSuspension()` runs, so this is a new option plus one line next to the pinned / audible checks: no new async calls, negligible per-suspend cost. It is what @arthurgbranco and @AmineI are asking for, and it also covers the monitoring / dashboard case (put those tabs in a group). @AmineI's tab-group-sync reason is a good one: suspended tabs not syncing is a real gap that this closes for grouped tabs.

**Per-group "never suspend this group" as a follow-up.** The store itself fits fine, same shape as the whitelist (a list of exclusions kept as an option, resolved at check time). The one real cost is identifying "the same group" across restarts: numeric group IDs are not stable, so it keys on title + color, which means either a `chrome.tabGroups.get()` per check or a small cached `groupId -> group` map invalidated on `tabGroups.onUpdated` / `onRemoved`. Moderate, not hard, but its own little subsystem, so it should not hold up the global toggle.

**On the composite key (title + color + member-URL fingerprint):** I would drop that. Keeping a URL fingerprint current means maintaining a persistent per-group index synced across `tabs.onUpdated` / `onCreated` / `onRemoved` / `onMoved` and `tabGroups.onUpdated`, which is a lot of always-on machinery for a niche exclusion, and it cuts against how the rest of TMS works (lazy, event-driven, nothing maintains an eager cross-tab index). The only thing it buys is disambiguating untitled groups, and the simpler answer there is to enable the "Never suspend this group" menu item only for named groups, disabled with a tooltip otherwise. Two groups with the same name and colour colliding is an acceptable edge case.

Order: #483 -> global grouped-tabs toggle -> per-group override. All targeting 9.1.0 if they land clean; the per-group one can slip to 9.2.0 if it needs more room.

**Comment by jadshaker** [comment]
I have opened draft PRs #489 and #490 but I haven't reviewed them yet. @gioxx I'll ping you here once they are ready.

**Comment by jadshaker** [comment]
#489 ready for review

**Comment by jadshaker** [comment]
Hi @gioxx 
For the changelog of the previous merged PR, they were under the 9.0.3 section, I think we should move it to 9.1.0, what do you think?

**Comment by jadshaker** [comment]
#489 ready for merging , i have resolved the conflict with the `master` branch

**Comment by jadshaker** [comment]
#490 ready for review. I think we can close this issue after merging both #489 and #490

**Comment by gioxx** [comment]
All three sub-asks from this thread are shipped now:
- "Suspend/unsuspend all tabs in a tab group" and its ungrouped counterpart (both merged)
- The global "never suspend tabs in a tab group" toggle (#489, merged)
- The per-group "never suspend this group" override (#490, merged, after nine review rounds and a good amount of hardening along the way)

Thanks to @mherkazandjian for kicking this off and @jadshaker for seeing the rest of it through, first contribution on the project and it turned into a genuinely solid piece of work. Closing this tracking issue; a handful of smaller follow-ups from the review process are tracked separately in #501 and #496/#499.
